### PR TITLE
Theme selection

### PR DESCRIPTION
**Pull request by Dune-jr**
image

Implemented a Theme selector in Settings/General.
Themes are located in `datasrc/ui/themes`. This is the content of my `themes` folder:

image

For each theme, you need at least one map, at most two and a 2:1 icon:
- `<name>_day.map`
- `<name>_night.map`
- `<name>.png` (icon, optional)

The (none) option lets you go back to the default checkboard. This overlaps with the menu checkbox to disable backgrounds, so it was removed.

The icon in the example is experimental and will probably change. I did not commit it.
Credits to @Sonix- for some great work and much appreciated help.

### CODE REVIEW

**Comment by oy** [comment]
I would make the name convention that way:
if you want to have the day/night cycle, you have to put a <name>_day and <name>_night map in there. Otherwise you put one with just <name> in there.
Would just use 1 icon for all maps and just render it for list entries that have a night/day pair.

File: src/game/client/components/maplayers.h [hunk 1]
```diff
@@ -20,8 +20,8 @@ class CMapLayers : public CComponent
 
 	static void EnvelopeEval(float TimeOffset, int Env, float *pChannels, void *pUser);
 
-	void LoadBackgroundMap();
 	void LoadEnvPoints(const CLayers *pLayers, array<CEnvPoint>& lEnvPoints);
+	void LoadBackgroundMap(bool TryBoth);
```
Comment (oy) [inline src/game/client/components/maplayers.h]: tries both everytime, parameter could be removed

File: src/game/client/components/menus_settings.cpp [hunk 2]
```diff
@@ -566,6 +566,90 @@ class CLanguage
...
+	else if(l > 6 && str_comp(aFullName+l-6, "_night") == 0)
+	{
+		str_copy(aThemeName, pName, min((int)sizeof(aThemeName),l-5));		
+		isNight = true;
+	}
+	else
+		return 0;
+	
+	// try to edit an existing theme
+	for(sorted_array<CTheme>::range r = pSelf->m_lThemes.all(); !r.empty(); r.pop_front()) // bit slow but whatever
```
Comment (oy) [inline src/game/client/components/menus_settings.cpp]: for(int i = 0; i < pSelf->m_lThemes.size(), ++i)
{

	if(str_comp(pSelf->m_lThemes[i].m_aName, aThemeName) == 0)

	{...

**Comment by Dune-jr** [comment]
> I would make the name convention that way:
> if you want to have the day/night cycle, you have to put a _day and _night map in there. Otherwise you put one with just in there.

So what happens if there is only `jungle_day.map`? All of a sudden the theme isn't `jungle` anymore, but `jungle_day`?

> Would just use 1 icon for all maps and just render it for list entries that have a night/day pair.

There is only one icon per theme right now. It looks like two because the .png I got was like that, it should be redone.

**Comment by oy** [comment]
> 
> 
> > I would make the name convention that way:
> > if you want to have the day/night cycle, you have to put a _day and _night map in there. Otherwise you put one with just in there.
> 
> So what happens if there is only `jungle_day.map`? All of a sudden the theme isn't `jungle` anymore, but `jungle_day`?

Either skip these cause of wrong convention or add them like it would be just "jungle" (in a special case when there is "jungle" and "jungle_day" in there, it will always load with "jungle").
> 
> > Would just use 1 icon for all maps and just render it for list entries that have a night/day pair.
> 
> There is only one icon per theme right now. It looks like two because the .png I got was like that, it should be redone.
Would just use 1 icon in general to indicate day/night cycle. Does it really need an individual icon for each theme to show that?

**Comment by Dune-jr** [comment]
> Either skip these cause of wrong convention or add them like it would be just "jungle" (in a special case when there is "jungle" and "jungle_day" in there, it will always load with "jungle").

Alright, will do.

> Would just use 1 icon in general to indicate day/night cycle. Does it really need an individual icon for each theme to show that?

The icon was not intended to indicate day/night cycle, but instead some preview of the map. e.g. a sky for heavens, a checkerboard for (none), etc. The one on the screenshot is more of a placeholder :/

If the icon should instead only indicate day/night cycles, then we can just load one, yeah.

**Comment by oy** [comment]
> > Would just use 1 icon in general to indicate day/night cycle. Does it really need an individual icon for each theme to show that?
> 
> The icon was not intended to indicate day/night cycle, but instead some preview of the map. e.g. a sky for heavens, a checkerboard for (none), etc. The one on the screenshot is more of a placeholder :/
> 
> If the icon should instead only indicate day/night cycles, then we can just load one, yeah.

Ah, ok.

**Comment by Dune-jr** [comment]
I get this (still a placeholder icon)
image

with this:

image

**Comment by oy** [comment]
Yeah, looks good.
Day/night indicator could be added to the icon graphic itself - left half bright, right side dark f.e.?

**Comment by Dune-jr** [comment]
> Yeah, looks good.
> Day/night indicator could be added to the icon graphic itself - left half bright, right side dark f.e.?

Yeah, we could do that. The icon being shaped 2:1 helps nicely.

**Comment by 1stDecadeTW** [comment]
@Dune-jr  Jungle_night isn't showing in the menu.

**Comment by Dune-jr** [comment]
@1stDecadeTW `jungle_night` is used by the Theme `jungle`, when you play in the evening. It is intended for `jungle_day` and `jungle_night` to be shown at once.

**Pull request merged**